docs/logging.asciidoc

```diff
@@ -0,0 +1,135 @@
+logging.getLogger("").addHandler(console)
+----
+
+
+[[structlog]]
+==== `structlog`
+
+We provide a http://www.structlog.org/en/stable/processors.html[processor] for
+http://www.structlog.org/en/stable/[`structlog`] which will add three new keys
+to the event_dict of any processed event:
+
+* `"transaction.id"`
+* `"trace.id"`
+* `"span.id"`
+
+[source,python]
+----
+from structlog import PrintLogger, wrap_logger
+from elasticapm.handlers.structlog import structlog_processor
+
+wrapped_logger = PrintLogger()
+logger = wrap_logger(wrapped_logger, processors=[structlog_processor])
+log = logger.new()
+log.msg("some_event")
+----
+
+
+[[log-correlation]]
+=== Log Correlation in Elasticsearch
+
+In order to correlate logs from your app with transactions captured by the
+Elastic APM Python Agent, your logs must contain one or more of the following
+identifiers:
+
+* transaction.id
+* trace.id
+* span.id
+
+If you're using structured logging, either https://docs.python.org/3/howto/logging-cookbook.html#implementing-structured-logging[with a custom solution]
+or with http://www.structlog.org/en/stable/[structlog (recommended)], then this
+is fairly easy. Throw the http://www.structlog.org/en/stable/api.html#structlog.processors.JSONRenderer[JSONRenderer]
+in, and you can easily use https://www.elastic.co/blog/structured-logging-filebeat[Filebeat]
+to pull these logs into Elasticsearch.
+
+However, without structured logging the task gets a little trickier. Here we
+recommend first making sure your LogRecord objects have the elasticapm
+attributes (see <<logging>>), and then you'll want to combine some specific
+formatting with a Grok pattern, either in Elasticsearch using
+https://www.elastic.co/guide/en/elasticsearch/reference/current/grok-processor.html[the grok processor],
+or in https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html[logstash with a plugin].
+
+Say you have a https://docs.python.org/3/library/logging.html#logging.Formatter[Formatter]
+that looks like this:
+
+[source,python]
+----
+import logging
+
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+----
+
+You can add the APM identifiers in an easy-to-parse manner:
+
+[source,python]
+----
+import logging
+
+format_string = (
+    "%(asctime)s - %(name)s - %(levelname)s - %(message)s "
+    "| elasticapm "
+    "transaction.id=%(elasticapm_transaction_id) "
+    "trace.id=%(elasticapm_trace_id) "
+    "span.id=%(elasticapm_span_id)"
+)
+formatter = logging.Formatter(format_string)
+----
+
+Then, you could use a grok pattern like this (for the
+https://www.elastic.co/guide/en/elasticsearch/reference/current/grok-processor.html[Elasticsearch Grok Processor]):
+
+
+[source, json]
+----
+{
+  "description" : "...",
+  "processors": [
+    {
+      "grok": {
+        "field": "message",
+        "patterns": ["%{GREEDYDATA:msg} | elasticapm transaction.id=%{DATA:transaction.id} trace.id=%{DATA:trace.id} span.id=%{DATA:span.id}"]
```

Comment:
This is why I made sure to use [DATA](https://github.com/elastic/logstash/blob/v1.4.2/patterns/grok-patterns#L14) instead of [WORD](https://github.com/elastic/logstash/blob/v1.4.2/patterns/grok-patterns#L11) or similar, because it's `.*` instead of `.+`. I think it should just resolve to an empty string for `None`. (Note, I always set the attribute, just set it to `None` if there is not transaction/trace/span ID.)

Reply:
Ah, gotcha. I'll retract my comment then 😊

Regarding missing attributes, I had the case in mind when a user configures the formatter, but not the filter or factory.

Reply:
Ah, I see. Yeah, because I always set it, we should be good. Thanks for asking!